### ISSUE TITLE
Sleep-wake works intermittently

### ISSUE THREAD

**Issue by mondalaci** [UltimateHackingKeyboard/firmware#1010]
When I put my PC to sleep with my UHK, the LEDs are disabled as expected. I press a key to wake up my PC, but then the LEDs are not enabled, and my UHK gets powerMode-locked for no reason. When I powerMode unlock it, the LEDs get enabled.

Then I put my PC to sleep again with my UHK, and the LEDs are not disabled now. I can't wake up my PC with a key, only with its power button.

The above two cases repeat forever when reproducing the sleep-wake cycle.

**Comment by kareltucek** [comment]
@benedekkupper can you make some sense of it?

TL;DR:
- the power delegate callbacks sometimes arrive, sometimes they don't. It seems like remote wakeup interferes with whether they arrive or not
- uhk80 can cause either the computer or all other peripherals freeze when it tries to send scancodes while the host is asleep

Edit: I have boldened those things that are wrong. 

---------------

Testing with current origin/master and linux. The master now contains contains:

- the power delegate solutiion
- power levels are logged onto the oled display straight from the power delegate
- there is some additional logic which turns the sleep mode on and off
- there is a modification which makes sure that oled is always avakened when logging onto the oled display

All testing is done while both uhk80 halves are connected to the computer, left one via hub, right one directly. Both batteries are connected. Some debugging probes are connected to the computer via a hub, but not to either UHK half.

I have tested the last two scenarios while only uhk80 right half was connected to the computer via usb, and observed the same results. (If needed, I can retest everything while only uhk80 right is connected)

Halves communicate over bluetooth.

## Testing scenarios:

### Test logging

- Add a Log("key down") so that it is triggered whenever a key is pressed; and comment out remote wakeup
- put UHK to sleep (either suspend computer, or invoke `powerMode sleep` macro)
- wait till oled goes dark
- tap a key
- observe oled going alive and logging "key down"
- conclusion: oled logging works for future scenarios
- (revert the code changes from the first step)

### Laszlo's scenario:

- suspend the computer
- computer suspends, UHK leds go dark, oled logs:

```
    L0 ON
    L2 Suspend
```
- I tap a key
- Computer resumes, but **UHK stays dark and logs nothing**
- I wake up UHK via a macro (`powerMode wake`); (This is a workaround. UHK should have woken up itself based on the usb events in the previous step.)
- UHK lits up and starts functioning again
- I suspend the computer again
- **UHK stays lit and logs nothing**
- I tap a key. (Since UHK doesn't know that computer is asleep, it attempts to send a scancode instead of waking the host up.)
- Computer doesn't wake up. (as expected)
- I wake the computer via computer's power button.
- Computer wakes up, **but *all* my keyboards freeze**. No input (mouse or keys) is accepted by the computer. Not from svalboard, not from laptop keyboard, not from ploopy, not from uhk60 and not from uhk80.
- Disconnect uhk80 from the usb (it is connected directly into the laptop)
- Now all the peripherals start working again

### Simple sleep / wake testing:
- I put computer to sleep using another keyboard
- uhk80 goes dark and logs:
```
    L0 ON
    L2 Suspend
```
- I wake up the computer up using its power button.
- computer screen lights up
- there is a few seconds delay
- now uhk80 lits up, starts working fine again and logs:
```
    L0 ON
```

### Sending scancodes while asleep

- disable sleep mode calls, disable remote wakeup calls
- put the computer to sleep
- (it goes to sleep allright, uhk is still lit as expected)
- tap a few keys (thus sending scancodes to the computer)
- wake the computer up
- now **again all peripherals (keyboards, mouses) are frozen**
- disconnect uhk80 from the usb
- (uhk80 is still battery powered here)
- now all peripherals work again
- connect uhk80 again
- **uhk80 is still frozen**, but other peripherals work
- use laptop keyboard to put the computer to sleep again (systemctl suspend)
- **now everything freezes again. The computer doesn't fall asleep.**
- disconnect uhk80 from usb
- now the computer puts itself to sleep


### Sending scancodes while asleep without semaphores

Some time ago I was adding some semaphores into the usb code and was wondering if I may have caused a deadlock, or some other issue with usb due to delays.

- disable sleep mode calls, disable remote wakeup calls, remove all k_sem* code from usb/*_app.cpp.
- retest previous scenario
- still the same results

**Comment by mondalaci** [comment]
This has been resolved.
